Embed instrument viewer in refl gui

gemmaguest opened the pull request:
**Description of work.**

This PR embeds the instrument viewer display in the preview tab on the ISIS Reflectometry GUI.
- An `InstViewModel` has been created to handle creating the instrument actor for the loaded workspace, and the surface to plot
- The `PreviewPresenter` has handles both the synchronous case - where the workspace already exists in the ADS and can be plotted immediately - and the async case where it needs to be loaded via an algorithm.
- `InstrumentActor` has been modified to take a message handler to handle pop-up messages rather than using QMessageBox itself. This is to avoid pop-up dialogs appearing when we are using `InstrumentActor` in unit tests.
- A change to `GLDisplay` was required to ensure the instrument view plot is blank when first created, rather than just rendering what was on screen previously

**To test:**

- Open the ISIS Reflectometry interface and set instrument to INTER
- Go to the Preview tab. The instrument viewer plot should be blank
- Enter a run number for an unloaded workspace e.g. `13460` and press enter or click Load
- The workspace should be displayed in the instrument plot
- Load a different workspace into mantid separately and give it a different name e.g. INTER45455
- Enter this name in the Preview tab and click load. This should use the workspace from the ADS rather than re-loading it. It should be displayed in the instrument view plot.

- Also test the basic functionality of instrument viewer to check that nothing is broken

Fixes #32490 

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

gemmaguest committed:
Add delegating constructor to InstrumentActor

Allow construction using a workspace or std::string.
Remove use of QString from public interface.
Minor tidying to use make_unique instead of new.

Re #32490

gemmaguest committed:
Add functions to create instrument view surface

- Create an instrument view model to use from PreviewPresenter to handle
creating the instrument view actor and surface required for plotting

Re #32490

gemmaguest committed:
Refactor PreviewPresenter dependencies into a struct

Re #32490

gemmaguest committed:
Add instrument view model to PreviewPresenter

Re #32490

DavidFair committed:
Add test for PreviewPresenter::notifyLoadCompleted

Adds the test implementation that gets / forwards the surface associated
when the workspace is loaded

Re #32490

DavidFair committed:
Wire up inst display in PreviewView

Wires up the inst display in the PreviewView so we can render an
instrument

Re #32490

DavidFair committed:
Fix empty render layer on GLDisplay

Fixes our GLDisplay having an empty buffer, this causes it to render the
last thing in video memory.
The existing code prepped the next buffer using the GL API, but did not
rotate it.

Re #32490

gemmaguest committed:
Plot result of loading workspace from ADS

Previously when a workspace existed in the ADS it was not hooked up to
actually continue to perform the plot. This commit refactors the code
slightly so that the presenter can call load from ADS first and plot the
result immediately if it succeeds; otherwise it starts loading async and
waits for a callback before doing the plotting, as per the existing
implementation.

Re #32490

gemmaguest committed:
Remove direct use of QMessageBox from InstrumentActor

Add a MessageHandler class to display messages and inject it into instrument actor

Re #32490

gemmaguest committed:
Rename reflectometry message handler

Improve clarity and fix compiler error due to name clash

Re #32490

gemmaguest committed:
Make refl message handler inherit from IMessageHandler

Re #32490

gemmaguest committed:
Inject message handler into InstViewModel

Re #32490

gemmaguest committed:
Move ownership of message handler out of instrument actor

Instrument actor needs to be recreated every time the workspace changes
so it makes passing ownership of dependencies to it problematic. Instead
we now pass a reference and the parent component is responsible for
ownership of the message handler

Re #32490

gemmaguest committed:
Remove temporary comments and logging

Re #32490

gemmaguest committed:
Minor fix to formatting to pass clang-format

This got messed up during a rebase

Re #32490

gemmaguest committed:
Fix linker error on Windows

Re #32490

gemmaguest commented:
@DavidFair I think this is ready for review if builds pass. I'm not sure why we made the change to QtTargetFunctions.cmake - was that deliberate? Also I'm not sure how to test the pop-ups still work in instrument viewer - any ideas?

DavidFair commented:
> QtTargetFunctions.cmake - was that deliberate? 
Yes, as we should make the downstream libraries we require to build a target public. Currently everyone is duplicating the list of deps each time they link a new target

> Also I'm not sure how to test the pop-ups still work in instrument viewer - any ideas?
We could test at least one by creating a sample workspace without an instrument. The others are much harder to trigger but as they all dispatch to the same concrete impl it should be "good enough" IMO

> Also I've just noticed that with INTER13460 I get a pop-up warning on the refl gui but not on the instrument viewer itself, which seems odd!
I'm getting this locally too, I'm not sure what's changed if something? I thought it was something upstream but it looks like nobody has touched the surrounding code for a while :confused:

gemmaguest committed:
Change instrument actor to run algorithms without the ADS

When using instrument actor from another GUI we are now using workspace
pointers rather than storing workspaces in the ADS so the instrument
actor needs to support this. As part of this we need to store a
non-const shared_ptr to the workspace rather than a weak_ptr. This makes
more sense anyway because the instrument actor modifies the workspace
e.g. when doing masking, it was just getting around this before by using
the ADS.

Re #32490

DavidFair commented:
This still isn't 100%, apply to workspace produces the error:
```
Error in execution of algorithm MaskDetectors
Add Data Object with empty name
```

Despite the fact it works correctly

DavidFair commented on qt/scientific_interfaces/ISISReflectometry/GUI/Preview/InstViewModel.cpp:
```diff
@@ -0,0 +1,47 @@
+}
+
+void InstViewModel::notifyWorkspaceUpdated(Mantid::API::MatrixWorkspace_sptr &workspace) {
+  // TODO refactor the component info stuff into the surface constructor so we don't need to get it here
```
Is this TODO still something we're planning on doing?

gemmaguest replied:
It would be good to do longer term but I think for expediency we shouldn't worry for now, so I think it best to leave the TODO in

DavidFair committed:
Fix newly detected cppcheck warnings

Fixes new warnings from extra checks that were included upstream, since
touching these files has broken the line suppressions
Re #32490

DavidFair committed:
Remove dead Qt4 code

Whilst fixing cppcheck warnings we can remove a load of completely dead
code, since the project does not build with Qt 4 < for a while now.
Re #32490

DavidFair commented:
@gemmaguest I'm happy to approve this PR if you want to review my last two commits?

gemmaguest commented:
> 
> 
> @gemmaguest I'm happy to approve this PR if you want to review my last two commits?

Yes, looks good to me :+1:

gemmaguest commented:
@mantidproject/gatekeepers Could we get this merged as it's blocking another PR?

martyngigg commented on qt/widgets/instrumentview/src/GLDisplay.cpp:
```diff
@@ -74,11 +74,19 @@ void GLDisplay::initializeGL() {
                1.0);
+  // Since we're manually manipulating the current GL buffer (as we're double buffered)
+  // we need to ask Qt to redraw, else we get a corrupted display with an empty surface
+  updateGL();
```
Nice. I think you've cleared up a long-standing bug with the display being empty before the view is manipulated in some cases.

martyngigg commented on qt/widgets/instrumentview/src/InstrumentActor.cpp:
```diff
@@ -5,19 +5,16 @@
 //   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
 // SPDX - License - Identifier: GPL - 3.0 +
 #include "MantidQtWidgets/InstrumentView/InstrumentActor.h"
-#if QT_VERSION < QT_VERSION_CHECK(5, 0, 0)
```
Thank you for removing the dead Qt4 code.
